File: 2024/haskell/day15a/Main.hs
module Main ( main ) where

import Data.List ( (\\), union )
import Data.Map.Strict ( Map )
import qualified Data.Map.Strict as Map
import Data.Set ( Set )
import qualified Data.Set as Set
import System.Environment ( getArgs, getProgName )
import System.Exit ( exitFailure )

type Position = (Int, Int)

type Direction = (Int, Int)

usage :: IO ()
usage = do
    progname <- getProgName
    putStrLn $ "usage: " ++ progname ++ " <file>"
    exitFailure

directionsMap :: Map Char Direction
directionsMap = Map.fromList [('^', (-1, 0)), ('v', (1, 0)), ('<', (0, -1)), ('>', (0, 1))]

scanGrid :: [String] -> ((Int, Int), Position, Set Position, Set Position)
scanGrid grid =
    let height = length grid
        width = length (head grid)
        robot = head [(r, c) | (r, row) <- zip [0..] grid,
                               (c, cell) <- zip [0..] row,
                               cell == '@']
        boxes = Set.fromList [(r, c) | (r, row) <- zip [0..] grid,
                                       (c, cell) <- zip [0..] row,
                                       cell == 'O']
        obstacles = Set.fromList [(r, c) | (r, row) <- zip [0..] grid,
                                           (c, cell) <- zip [0..] row,
                                           cell == '#']
    in ((height, width), robot, boxes, obstacles)

splitInput :: [String] -> ([String], String)
splitInput contentLines = (takeWhile (/= "") contentLines, filter (/= '\n') $ unlines $ tail $ dropWhile (/= "") contentLines)

moveRobot :: Char -> Position -> Set Position -> Set Position -> (Position, Set Position)
moveRobot dir robot@(x, y) boxes obstacles =
    let (dx, dy) = case Map.lookup dir directionsMap of
                        Just delta -> delta
                        Nothing -> error ("unexpected character in input: " ++ show dir)
        tryMove (dx, dy) (rx, ry) movedBoxes =
            if null movedBoxes
                then let (nx, ny) = (rx + dx, ry + dy)
                     in if (nx, ny) `Set.notMember` boxes
                            then if (nx, ny) `Set.notMember` obstacles
                                    then ((nx, ny), boxes)
                                    else ((rx, ry), boxes)
                            else tryMove (dx, dy) (rx, ry) ((nx, ny):movedBoxes)
                else let (bx, by) = head movedBoxes
                         (nx, ny) = (bx + dx, by + dy)
                     in if (nx, ny) `Set.notMember` boxes
                            then if (nx, ny) `Set.notMember` obstacles
                                    then let tmpBoxes = Set.toList boxes \\ movedBoxes
                                             newPositions = [(px + dx, py + dy) | (px, py) <- movedBoxes]
                                         in ((rx + dx, ry + dy), Set.fromList (tmpBoxes `union` newPositions))
                                    else ((rx, ry), boxes)
                            else tryMove (dx, dy) (rx, ry) ((nx, ny):movedBoxes)
    in tryMove (dx, dy) (x, y) []

processMoves :: String -> Position -> Set Position -> Set Position -> (Position, Set Position)
processMoves [] robot boxes _ = (robot, boxes)
processMoves (dir:rest) robot boxes obstacles =
    let (robot', boxes') = moveRobot dir robot boxes obstacles
    in processMoves rest robot' boxes' obstacles

calcGPS :: Position -> Int
calcGPS (r, c) = (100 * r) + c

process :: String -> Int
process contents =
    let (grid, directions) = splitInput (lines contents)
        ((height, width), robot, boxes, obstacles) = scanGrid grid
        (_, finalBoxes) = processMoves directions robot boxes obstacles
    in sum $ map calcGPS $ Set.toList finalBoxes

main :: IO ()
main = do
    args <- getArgs
    case args of
        [filename] -> do
            contents <- readFile filename
            let result = process contents
            putStrLn $ "result = " ++ show result
        _ -> usage

File: 2024/rust/day15a/src/main.rs
use std::collections::{HashSet, VecDeque};
use std::env;
use std::fs;
use std::process;
use std::time::SystemTime;

type Position = (u32, u32);

#[derive(Clone)]
struct Warehouse {
    dimensions: (u32, u32),
    robot: Position,
    boxes: HashSet<Position>,
    obstacles: HashSet<Position>,
}

fn usage() {
    let progname = env::args().next().unwrap();
    println!("usage: {progname} <file>");
    process::exit(1);
}

fn scan_input(contents: &str) -> (Warehouse, Vec<char>) {
    let mut moves: Vec<char> = vec![];
    let mut obstacles: HashSet<Position> = HashSet::new();
    let mut boxes: HashSet<Position> = HashSet::new();
    let mut robot: Position = (0, 0);
    let mut dimensions: (u32, u32) = (0, 0);
    let mut horiz = 0;
    for (row, line) in contents.lines().enumerate() {
        if line.starts_with("#") {
            if line.chars().all(|x| x == '#') {
                horiz += 1;
            }
            for (col, ch) in line.chars().enumerate() {
                match ch {
                    '#' => {
                        obstacles.insert((row as u32, col as u32));
                    }
                    'O' => {
                        boxes.insert((row as u32, col as u32));
                    }
                    '@' => robot = (row as u32, col as u32),
                    _ => (),
                }
            }
            if horiz == 2 {
                dimensions = ((row + 1) as u32, line.chars().count() as u32);
            }
        } else if line.starts_with(['^', 'v', '<', '>']) {
            moves.append(&mut line.chars().collect::<Vec<char>>());
        }
    }
    (
        Warehouse {
            dimensions,
            robot,
            boxes,
            obstacles,
        },
        moves,
    )
}

fn move_robot(warehouse: &Warehouse, dir: (i32, i32)) -> Warehouse {
    let next_robot_pos = (
        (warehouse.robot.0 as i32 + dir.0) as u32,
        (warehouse.robot.1 as i32 + dir.1) as u32,
    );
    if warehouse.obstacles.contains(&next_robot_pos) {
        warehouse.clone()
    } else if warehouse.boxes.contains(&next_robot_pos) {
        let mut boxes_to_move: HashSet<Position> = HashSet::new();
        let initial_box = next_robot_pos;
        let mut q: VecDeque<Position> = VecDeque::new();
        q.push_back(initial_box);
        while let Some(curr_box) = q.pop_front() {
            boxes_to_move.insert(curr_box);
            let next_box = (
                (curr_box.0 as i32 + dir.0) as u32,
                (curr_box.1 as i32 + dir.1) as u32,
            );
            if warehouse.boxes.contains(&next_box) {
                q.push_back(next_box);
            }
        }
        let mut working_set: HashSet<Position> = HashSet::new();
        for pos in boxes_to_move.iter() {
            working_set.insert(((pos.0 as i32 + dir.0) as u32, (pos.1 as i32 + dir.1) as u32));
        }
        if warehouse.obstacles.intersection(&working_set).count() > 0 {
            return warehouse.clone();
        } else {
            let stationary_boxes: HashSet<Position> = warehouse
                .boxes
                .difference(&boxes_to_move)
                .cloned()
                .collect();
            let moved_boxes: HashSet<Position> = boxes_to_move
                .iter()
                .map(|pos| ((pos.0 as i32 + dir.0) as u32, (pos.1 as i32 + dir.1) as u32))
                .collect();
            let new_boxes: HashSet<Position> =
                stationary_boxes.union(&moved_boxes).cloned().collect();
            return Warehouse {
                dimensions: warehouse.dimensions,
                robot: next_robot_pos,
                boxes: new_boxes,
                obstacles: warehouse.obstacles.clone(),
            };
        }
    } else {
        return Warehouse {
            dimensions: warehouse.dimensions,
            robot: next_robot_pos,
            boxes: warehouse.boxes.clone(),
            obstacles: warehouse.obstacles.clone(),
        };
    }
}

fn process_moves(warehouse: &Warehouse, moves: &[char]) -> Warehouse {
    let mut new_warehouse = Warehouse {
        dimensions: warehouse.dimensions,
        robot: warehouse.robot,
        boxes: warehouse.boxes.clone(),
        obstacles: warehouse.obstacles.clone(),
    };
    for curr_move in moves {
        let dir: (i32, i32) = match curr_move {
            '^' => (-1, 0),
            'v' => (1, 0),
            '<' => (0, -1),
            '>' => (0, 1),
            _ => panic!("invalid move"),
        };
        new_warehouse = move_robot(&new_warehouse, dir);
    }
    new_warehouse
}

fn process(contents: &str) -> u32 {
    let (warehouse, moves) = scan_input(contents);
    let final_state = process_moves(&warehouse, &moves);
    final_state
        .boxes
        .iter()
        .map(|pos| pos.0 * 100 + pos.1)
        .sum::<u32>()
}

fn calc_runtime(start_time: &SystemTime) -> String {
    let mut elapsed = start_time.elapsed().unwrap().as_micros() as f64;
    let mut units = "μs";
    if elapsed > 1000.0 {
        elapsed /= 1000.0;
        units = "ms";
    }
    if elapsed > 1000.0 {
        elapsed /= 1000.0;
        units = "s";
    }
    format!("(elapsed time: {elapsed} {units})")
}

fn main() {
    let start_time = SystemTime::now();
    if env::args().count() < 2 {
        usage();
    }
    let filename = env::args().nth(1).unwrap();
    let contents = fs::read_to_string(filename).expect("read of input file failed");
    let result = process(&contents);
    let duration = calc_runtime(&start_time);
    println!("result = {result}  {duration}");
}
